### PR TITLE
perf(rust, python): Rolling min/max with nulls

### PR DESCRIPTION
**Pull request by magarick**
Brings the rolling extremum improvements for sequences with partially sorted runs to the case where null values are present.

Code to generate benchmarking data and performance tables below. I tested on series with 0.1% random nulls (lo), ~40% random nulls (hi), and ~2.25% nulls as one long run in the middle (run). Timings are for 50 runs.

It's substantially faster on data with long runs (p2) as well as for a minimum on ascending data (it turns out this special case was missing). It also performs better in dynamic groupbys. Other than that, its performance is similar when using "normal" rolling windows that increment by one. It's a little slower for short window lengths and I don't know how to shrink this gap, or if it's possible, but it's small and worth the other improvements.

<details>
<summary>Data generation code</summary>

```python
ds = pl.date_range(datetime(2013, 1, 1), datetime(2022, 12, 31), '1h', eager = True).alias('ds')
def generate_data(ds, null_idxs, N = 50):
    np.random.seed(100)
    cols = [pl.Series('d', ds.to_numpy().repeat(N)),
        pl.Series('v', np.random.standard_normal(N * len(ds))),
        pl.Series('x', np.random.randint(0, 50, len(ds)).repeat(N)),
        pl.Series('s', range(len(ds) * N)),
        pl.Series('r', range(len(ds) * N)).reverse(),
        pl.Series('p', (list(range(125)) + list(range(125, 0, -1))) * (701 * N//2)),
        pl.Series('p2', (list(range(17525)) + list(range(17525, 0, -1))) * (5 * N//2))
    ]
    for c in cols[1:]:
        c.set_at_idx(null_idxs, None)
    dat = pl.DataFrame(cols)
    return dat.set_sorted('d')

np.random.seed(100)
nulls = {
    'lo': np.random.choice(ds.len() * N, round(0.001 * ds.len() * N)),
    # I meant to sample without replacement here, oops.
    # But this still makes about 40% of the values null. Good enough
    'hi': np.random.choice(ds.len() * N, round(0.5 * ds.len() * N)),
    'run': range(100000, 200000)
}

dats = {k: generate_data(ds, v) for k, v in nulls.items()}
```

</details>

<details>
<summary>Rolling by one performance</summary>

|nulls |variable | window|op  |  current|    new|
|:-----|:--------|------:|:---|--------:|------:|
|lo    |x        |      5|min |    3.233|  3.076|
|lo    |x        |      5|max |    3.082|  3.191|
|lo    |x        |    250|min |    3.894|  3.874|
|lo    |x        |    250|max |    3.747|  3.903|
|lo    |x        |   7500|min |    3.962|  3.071|
|lo    |x        |   7500|max |    3.729|  3.140|
|lo    |v        |      5|min |    4.642|  5.281|
|lo    |v        |      5|max |    4.626|  5.498|
|lo    |v        |    250|min |    4.205|  3.521|
|lo    |v        |    250|max |    4.184|  3.831|
|lo    |v        |   7500|min |    4.247|  3.556|
|lo    |v        |   7500|max |    4.173|  3.826|
|lo    |s        |      5|min |    5.916|  5.822|
|lo    |s        |      5|max |    3.175|  3.210|
|lo    |s        |    250|min |  127.216|  8.475|
|lo    |s        |    250|max |    3.191|  3.257|
|lo    |s        |   7500|min | 3670.713| 21.678|
|lo    |s        |   7500|max |    3.174|  3.362|
|lo    |r        |      5|min |    3.369|  3.313|
|lo    |r        |      5|max |    4.894|  5.837|
|lo    |r        |    250|min |    3.369|  3.328|
|lo    |r        |    250|max |    4.944|  8.561|
|lo    |r        |   7500|min |    3.438|  3.375|
|lo    |r        |   7500|max |    5.014| 21.732|
|lo    |p        |      5|min |    4.584|  4.556|
|lo    |p        |      5|max |    4.377|  4.559|
|lo    |p        |    250|min |    3.391|  3.090|
|lo    |p        |    250|max |    3.189|  3.086|
|lo    |p        |   7500|min |    3.474|  3.165|
|lo    |p        |   7500|max |    3.230|  3.229|
|lo    |p2       |      5|min |    4.632|  4.575|
|lo    |p2       |      5|max |    4.424|  4.558|
|lo    |p2       |    250|min |   64.875|  5.935|
|lo    |p2       |    250|max |   64.628|  5.926|
|lo    |p2       |   7500|min | 1447.343| 45.077|
|lo    |p2       |   7500|max | 1444.542| 44.799|
|hi    |x        |      5|min |    4.793|  4.735|
|hi    |x        |      5|max |    4.561|  4.883|
|hi    |x        |    250|min |    6.004|  5.457|
|hi    |x        |    250|max |    5.760|  5.466|
|hi    |x        |   7500|min |    6.111|  4.076|
|hi    |x        |   7500|max |    5.823|  4.054|
|hi    |v        |      5|min |    6.103|  6.588|
|hi    |v        |      5|max |    6.026|  6.676|
|hi    |v        |    250|min |    6.213|  4.969|
|hi    |v        |    250|max |    6.048|  5.207|
|hi    |v        |   7500|min |    6.128|  4.825|
|hi    |v        |   7500|max |    6.032|  5.120|
|hi    |s        |      5|min |    5.523|  7.220|
|hi    |s        |      5|max |    4.183|  4.242|
|hi    |s        |    250|min |   76.968|  8.740|
|hi    |s        |    250|max |    4.732|  4.169|
|hi    |s        |   7500|min | 4161.749| 17.031|
|hi    |s        |   7500|max |    4.726|  4.166|
|hi    |r        |      5|min |    4.404|  4.204|
|hi    |r        |      5|max |    6.528|  7.226|
|hi    |r        |    250|min |    5.032|  4.124|
|hi    |r        |    250|max |    6.434|  8.736|
|hi    |r        |   7500|min |    5.035|  4.121|
|hi    |r        |   7500|max |    6.504| 17.012|
|hi    |p        |      5|min |    5.194|  5.868|
|hi    |p        |      5|max |    4.969|  5.964|
|hi    |p        |    250|min |    5.455|  4.469|
|hi    |p        |    250|max |    5.155|  4.482|
|hi    |p        |   7500|min |    5.447|  4.144|
|hi    |p        |   7500|max |    5.144|  4.133|
|hi    |p2       |      5|min |    5.035|  5.779|
|hi    |p2       |      5|max |    4.803|  5.864|
|hi    |p2       |    250|min |   40.767|  6.497|
|hi    |p2       |    250|max |   40.589|  6.570|
|hi    |p2       |   7500|min | 1643.001| 52.595|
|hi    |p2       |   7500|max | 1630.685| 53.946|
|run   |x        |      5|min |    3.191|  3.061|
|run   |x        |      5|max |    3.048|  3.191|
|run   |x        |    250|min |    3.842|  3.809|
|run   |x        |    250|max |    3.669|  3.829|
|run   |x        |   7500|min |    3.894|  2.996|
|run   |x        |   7500|max |    3.629|  2.953|
|run   |v        |      5|min |    4.583|  5.224|
|run   |v        |      5|max |    4.546|  5.438|
|run   |v        |    250|min |    4.121|  3.501|
|run   |v        |    250|max |    4.079|  3.781|
|run   |v        |   7500|min |    4.050|  3.408|
|run   |v        |   7500|max |    3.977|  3.692|
|run   |s        |      5|min |    5.822|  5.744|
|run   |s        |      5|max |    3.118|  3.206|
|run   |s        |    250|min |  124.151|  8.335|
|run   |s        |    250|max |    3.111|  3.199|
|run   |s        |   7500|min | 3579.154| 21.089|
|run   |s        |   7500|max |    3.122|  3.198|
|run   |r        |      5|min |    3.421|  3.253|
|run   |r        |      5|max |    4.884|  5.776|
|run   |r        |    250|min |    3.416|  3.249|
|run   |r        |    250|max |    5.902|  8.411|
|run   |r        |   7500|min |    3.391|  3.263|
|run   |r        |   7500|max |   33.131| 21.218|
|run   |p        |      5|min |    4.528|  4.520|
|run   |p        |      5|max |    4.328|  4.530|
|run   |p        |    250|min |    3.438|  3.059|
|run   |p        |    250|max |    3.136|  3.033|
|run   |p        |   7500|min |    3.414|  3.068|
|run   |p        |   7500|max |    3.177|  3.048|
|run   |p2       |      5|min |    4.604|  4.522|
|run   |p2       |      5|max |    4.364|  4.515|
|run   |p2       |    250|min |   63.324|  5.814|
|run   |p2       |    250|max |   63.116|  5.822|
|run   |p2       |   7500|min | 1410.123| 43.794|
|run   |p2       |   7500|max | 1409.879| 43.512|

</details>

<details>
<summary>Dynamic groupby performance</summary>

|nulls |variable |window                    |op  | current|    new|
|:-----|:--------|:-------------------------|:---|-------:|------:|
|lo    |x        |(every = 1h; period = 2h) |min |   1.814|  1.264|
|lo    |x        |(every = 1h; period = 2h) |max |   1.818|  1.252|
|lo    |v        |(every = 1h; period = 2h) |min |   2.402|  1.852|
|lo    |v        |(every = 1h; period = 2h) |max |   2.521|  1.866|
|lo    |s        |(every = 1h; period = 2h) |min |   1.852|  1.444|
|lo    |s        |(every = 1h; period = 2h) |max |   1.935|  1.199|
|lo    |r        |(every = 1h; period = 2h) |min |   1.906|  1.199|
|lo    |r        |(every = 1h; period = 2h) |max |   1.825|  1.433|
|lo    |p        |(every = 1h; period = 2h) |min |   1.893|  1.318|
|lo    |p        |(every = 1h; period = 2h) |max |   1.835|  1.232|
|lo    |p2       |(every = 1h; period = 2h) |min |   1.859|  1.273|
|lo    |p2       |(every = 1h; period = 2h) |max |   1.865|  1.234|
|lo    |x        |(every = 1d; period = 5d) |min |   2.203|  1.262|
|lo    |x        |(every = 1d; period = 5d) |max |   2.210|  1.261|
|lo    |v        |(every = 1d; period = 5d) |min |   2.777|  1.749|
|lo    |v        |(every = 1d; period = 5d) |max |   2.872|  1.755|
|lo    |s        |(every = 1d; period = 5d) |min |   3.515|  1.258|
|lo    |s        |(every = 1d; period = 5d) |max |   2.125|  1.243|
|lo    |r        |(every = 1d; period = 5d) |min |   2.124|  1.250|
|lo    |r        |(every = 1d; period = 5d) |max |   3.512|  1.263|
|lo    |p        |(every = 1d; period = 5d) |min |   1.588|  1.252|
|lo    |p        |(every = 1d; period = 5d) |max |   1.600|  1.241|
|lo    |p2       |(every = 1d; period = 5d) |min |   2.723|  1.298|
|lo    |p2       |(every = 1d; period = 5d) |max |   2.730|  1.258|
|lo    |x        |(every = 1h; period = 5d) |min |  19.656| 12.760|
|lo    |x        |(every = 1h; period = 5d) |max |  19.659| 12.747|
|lo    |v        |(every = 1h; period = 5d) |min |  20.311| 13.512|
|lo    |v        |(every = 1h; period = 5d) |max |  20.418| 13.525|
|lo    |s        |(every = 1h; period = 5d) |min |  76.977| 13.164|
|lo    |s        |(every = 1h; period = 5d) |max |  19.193| 12.723|
|lo    |r        |(every = 1h; period = 5d) |min |  19.217| 12.728|
|lo    |r        |(every = 1h; period = 5d) |max |  77.020| 13.194|
|lo    |p        |(every = 1h; period = 5d) |min |  19.106| 12.816|
|lo    |p        |(every = 1h; period = 5d) |max |  19.160| 12.749|
|lo    |p2       |(every = 1h; period = 5d) |min |  43.235| 13.352|
|lo    |p2       |(every = 1h; period = 5d) |max |  43.235| 13.317|
|hi    |x        |(every = 1h; period = 2h) |min |   2.918|  1.959|
|hi    |x        |(every = 1h; period = 2h) |max |   2.921|  1.885|
|hi    |v        |(every = 1h; period = 2h) |min |   3.470|  2.535|
|hi    |v        |(every = 1h; period = 2h) |max |   3.552|  2.606|
|hi    |s        |(every = 1h; period = 2h) |min |   3.068|  1.987|
|hi    |s        |(every = 1h; period = 2h) |max |   3.104|  1.847|
|hi    |r        |(every = 1h; period = 2h) |min |   3.105|  1.873|
|hi    |r        |(every = 1h; period = 2h) |max |   3.068|  2.020|
|hi    |p        |(every = 1h; period = 2h) |min |   3.082|  2.141|
|hi    |p        |(every = 1h; period = 2h) |max |   3.016|  1.926|
|hi    |p2       |(every = 1h; period = 2h) |min |   3.085|  1.944|
|hi    |p2       |(every = 1h; period = 2h) |max |   3.093|  1.823|
|hi    |x        |(every = 1d; period = 5d) |min |   3.502|  1.985|
|hi    |x        |(every = 1d; period = 5d) |max |   3.516|  1.953|
|hi    |v        |(every = 1d; period = 5d) |min |   4.067|  2.555|
|hi    |v        |(every = 1d; period = 5d) |max |   4.151|  2.673|
|hi    |s        |(every = 1d; period = 5d) |min |   6.176|  1.732|
|hi    |s        |(every = 1d; period = 5d) |max |   3.309|  1.820|
|hi    |r        |(every = 1d; period = 5d) |min |   3.307|  1.824|
|hi    |r        |(every = 1d; period = 5d) |max |   6.200|  1.722|
|hi    |p        |(every = 1d; period = 5d) |min |   2.369|  1.914|
|hi    |p        |(every = 1d; period = 5d) |max |   2.389|  1.866|
|hi    |p2       |(every = 1d; period = 5d) |min |   4.553|  1.939|
|hi    |p2       |(every = 1d; period = 5d) |max |   4.564|  1.834|
|hi    |x        |(every = 1h; period = 5d) |min |  21.425| 13.562|
|hi    |x        |(every = 1h; period = 5d) |max |  21.431| 13.529|
|hi    |v        |(every = 1h; period = 5d) |min |  21.742| 14.473|
|hi    |v        |(every = 1h; period = 5d) |max |  21.855| 14.706|
|hi    |s        |(every = 1h; period = 5d) |min | 122.575| 13.857|
|hi    |s        |(every = 1h; period = 5d) |max |  20.407| 13.419|
|hi    |r        |(every = 1h; period = 5d) |min |  20.404| 13.391|
|hi    |r        |(every = 1h; period = 5d) |max | 123.651| 13.853|
|hi    |p        |(every = 1h; period = 5d) |min |  20.648| 13.496|
|hi    |p        |(every = 1h; period = 5d) |max |  20.718| 13.377|
|hi    |p2       |(every = 1h; period = 5d) |min |  62.987| 14.526|
|hi    |p2       |(every = 1h; period = 5d) |max |  63.431| 14.439|
|run   |x        |(every = 1h; period = 2h) |min |   1.778|  1.529|
|run   |x        |(every = 1h; period = 2h) |max |   1.798|  1.253|
|run   |v        |(every = 1h; period = 2h) |min |   2.355|  1.816|
|run   |v        |(every = 1h; period = 2h) |max |   2.471|  1.829|
|run   |s        |(every = 1h; period = 2h) |min |   1.827|  1.382|
|run   |s        |(every = 1h; period = 2h) |max |   1.965|  1.193|
|run   |r        |(every = 1h; period = 2h) |min |   1.970|  1.197|
|run   |r        |(every = 1h; period = 2h) |max |   1.933|  1.383|
|run   |p        |(every = 1h; period = 2h) |min |   1.902|  1.307|
|run   |p        |(every = 1h; period = 2h) |max |   2.067|  1.240|
|run   |p2       |(every = 1h; period = 2h) |min |   2.184|  1.271|
|run   |p2       |(every = 1h; period = 2h) |max |   1.997|  1.237|
|run   |x        |(every = 1d; period = 5d) |min |   2.384|  1.254|
|run   |x        |(every = 1d; period = 5d) |max |   2.278|  1.260|
|run   |v        |(every = 1d; period = 5d) |min |   2.770|  1.735|
|run   |v        |(every = 1d; period = 5d) |max |   2.927|  1.737|
|run   |s        |(every = 1d; period = 5d) |min |   3.539|  1.257|
|run   |s        |(every = 1d; period = 5d) |max |   2.408|  1.245|
|run   |r        |(every = 1d; period = 5d) |min |   2.211|  1.245|
|run   |r        |(every = 1d; period = 5d) |max |   3.697|  1.257|
|run   |p        |(every = 1d; period = 5d) |min |   1.684|  1.256|
|run   |p        |(every = 1d; period = 5d) |max |   1.596|  1.251|
|run   |p2       |(every = 1d; period = 5d) |min |   2.693|  1.312|
|run   |p2       |(every = 1d; period = 5d) |max |   2.696|  1.267|
|run   |x        |(every = 1h; period = 5d) |min |  19.626| 12.769|
|run   |x        |(every = 1h; period = 5d) |max |  19.630| 12.750|
|run   |v        |(every = 1h; period = 5d) |min |  20.259| 13.517|
|run   |v        |(every = 1h; period = 5d) |max |  20.368| 13.584|
|run   |s        |(every = 1h; period = 5d) |min |  75.429| 13.231|
|run   |s        |(every = 1h; period = 5d) |max |  19.185| 12.966|
|run   |r        |(every = 1h; period = 5d) |min |  19.172| 12.868|
|run   |r        |(every = 1h; period = 5d) |max |  75.443| 13.802|
|run   |p        |(every = 1h; period = 5d) |min |  19.068| 12.790|
|run   |p        |(every = 1h; period = 5d) |max |  19.128| 12.747|
|run   |p2       |(every = 1h; period = 5d) |min |  42.541| 13.385|
|run   |p2       |(every = 1h; period = 5d) |max |  42.602| 13.283|

</details>

### CODE REVIEW

File: polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs [hunk 1]
```diff
@@ -1,460 +1,247 @@
...
-
-        // return first non null
-        for idx in start..end {
-            let valid = self.validity.get_bit_unchecked(idx);
-
-            if valid {
-                return Some(*self.slice.get_unchecked(idx));
+// Identify runs of sorted values, either increasing or decreasing. It lets us avoid unnecessary
+// comparisons when the extremum keeps dropping off the front of the window
+macro_rules! fn_n_sorted_past {
```
Comment (ritchie46) [inline polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs:18]: Can we rewrite this macro with generics? I only want macro's to do dispatch and very small code things like `polars_err!`. 

The parts you want variable, you can put behind auxilliary traits. Happy to help if needed.

Comment (magarick) [reply polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs]: Ok, now it's slower. Even after I changed the code a bit to try and optimize it's better than my first attempt with  generics but still slower than the macro version. I'll take another crack at it but maybe I'm missing something obvious.

Comment (magarick) [reply polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs]: A little better with some const generic hackery. Tolerable now but still looks a little slower on average than the macro version.

Comment (magarick) [reply polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs]: Any thoughts? I can make it a tiny bit faster by making the compare part branchless now but I don't think it matters. Maybe there are other improvements too but it seems alright. Are you happy with this style overall?

Comment (magarick) [reply polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs]: What else do we need to do to get this landed. It fixes some pretty nasty edge cases and I'd like to put rolling min/max behind me for a while :-)

Comment (ritchie46) [reply polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs]: Didn't get to it yet. :) Its on my stack!

Comment (magarick) [reply polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs]: Ok, no problem. I know big-ish ones like this probably take more effort to review.

File: polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs [hunk 2]
```diff
@@ -1,460 +1,256 @@
...
-
-        // return first non null
-        for idx in start..end {
-            let valid = self.validity.get_bit_unchecked(idx);
-
-            if valid {
-                return Some(*self.slice.get_unchecked(idx));
+// R should be an Ordering but isn't because non-primitive const generics aren't considered stable
+// as of this writing. Passing comparison information this way is faster than having a field inside
+// the struct so we live with it.
```
Comment (ritchie46) [inline polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs:19]: Could you explain to me what you wanted to do without the `R` trick? 

I believe it should be solvable with an intermediate trait.


What was your initial solution that didn't work?

Comment (magarick) [reply polars/polars-arrow/src/kernels/rolling/nulls/min_max.rs:19]: I tried it a few different ways. One was to pass a comparison function in to the constructor and call that dynamically. That's what the previous version of the code did.
The other was to take the Ordering passed in and use that or it's `reverse()` to determine where `NaN` goes. Both were slower, though the second was a bit better.

The current solution seems roughly on par with macros.

If there's a way to do it by adding more traits I guess we can do that. In general, I'm loath to add such structure and complexity if it isn't absolutely needed. But this is pretty deep internals, it's localized, and I care much more about how the external facing parts look for end-users. Besides, if it's more verbose I'll get more GitHubPoints :-)

**Comment by stinodego** [comment]
I am closing this PR as stale.

We have a new rolling min/max algorithm in the works, so the work here will be obsolete.

**Pull request closed without merging**